File: packages/code-storage-typescript/tests/standard-vocabulary.test.ts
import { beforeEach, describe, expect, it, vi } from 'vitest';

import { GitStorage } from '../src/index';

type MockFetch = ReturnType<typeof vi.fn>;

const mockFetch = vi.fn() as MockFetch;
vi.stubGlobal('fetch', mockFetch);

function jsonResponse(body: unknown): Response {
  return new Response(JSON.stringify(body), {
    status: 200,
    headers: { 'content-type': 'application/json' },
  });
}

function makeRepo() {
  const store = new GitStorage({ name: 'v0', token: 'header.payload.signature' });
  return store.repo({ id: 'repo' });
}

function commitInfo() {
  return {
    sha: 'subject-sha',
    parent_shas: [],
    message: 'message',
    author_name: 'Author',
    author_email: 'author@example.com',
    committer_name: 'Committer',
    committer_email: 'committer@example.com',
    date: '2026-08-28T00:00:00Z',
  };
}

function diffResponse(extra: Record<string, unknown> = {}) {
  return {
    sha: 'subject-sha',
    stats: { files: 0, additions: 0, deletions: 0, changes: 0 },
    files: [],
    filtered_files: [],
    ...extra,
  };
}

function mergeResponse(source: Record<string, unknown>) {
  return {
    result: 'fast_forward',
    commit_sha: 'new-sha',
    tree_sha: 'tree-sha',
    source: { ...source, ephemeral: false, sha: 'source-sha' },
    target: {
      branch: 'main',
      ephemeral: false,
      old_sha: 'old-sha',
      new_sha: 'new-sha',
    },
    promoted_commits: 1,
  };
}

function noteResponse(refs: Record<string, unknown>) {
  return {
    sha: 'object-sha',
    ...refs,
    new_ref_sha: 'notes-sha',
    result: { success: true, status: 'ok' },
  };
}

function commitResponse(branches: Record<string, unknown>) {
  return {
    commit: {
      commit_sha: 'new-sha',
      tree_sha: 'tree-sha',
      target_branch: 'main',
      pack_bytes: 1,
      blob_count: 0,
    },
    result: {
      ...branches,
      old_sha: 'old-sha',
      new_sha: 'new-sha',
      success: true,
      status: 'ok',
    },
  };
}

async function readRequestBody(body: unknown): Promise<string> {
  if (!body) return '';
  if (typeof body === 'string') return body;
  if (body instanceof Uint8Array) return new TextDecoder().decode(body);
  if (
    typeof body === 'object' &&
    body !== null &&
    'getReader' in body
  ) {
    const reader = (body as ReadableStream<Uint8Array>).getReader();
    const chunks: Uint8Array[] = [];
    while (true) {
      const { value, done } = await reader.read();
      if (done) break;
      if (value) chunks.push(value);
    }
    const size = chunks.reduce((total, chunk) => total + chunk.byteLength, 0);
    const result = new Uint8Array(size);
    let offset = 0;
    for (const chunk of chunks) {
      result.set(chunk, offset);
      offset += chunk.byteLength;
    }
    return new TextDecoder().decode(result);
  }
  return '';
}

describe('standard API vocabulary', () => {
  beforeEach(() => {
    mockFetch.mockReset();
  });

  it('sends ref for getCommit and prefers it to sha', async () => {
    mockFetch.mockImplementationOnce((url) => {
      const params = new URL(String(url)).searchParams;
      expect(params.get('ref')).toBe('preferred');
      expect(params.has('sha')).toBe(false);
      return Promise.resolve(jsonResponse({ commit: commitInfo() }));
    });

    await makeRepo().getCommit({ ref: 'preferred', sha: 'legacy' });
  });

  it('sends ref for listCommits and prefers it to branch', async () => {
    mockFetch.mockImplementationOnce((url) => {
      const params = new URL(String(url)).searchParams;
      expect(params.get('ref')).toBe('preferred');
      expect(params.has('branch')).toBe(false);
      return Promise.resolve(
        jsonResponse({ commits: [], has_more: false })
      );
    });

    await makeRepo().listCommits({ ref: 'preferred', branch: 'legacy' });
  });

  it('sends standard commit diff query names', async () => {
    mockFetch.mockImplementationOnce((url) => {
      const params = new URL(String(url)).searchParams;
      expect(params.get('ref')).toBe('preferred');
      expect(params.get('base_ref')).toBe('preferred-base');
      expect(params.get('ref_is_ephemeral')).toBe('false');
      expect(params.get('base_is_ephemeral')).toBe('false');
      expect(params.get('git_apply_compatible')).toBe('true');
      expect(params.has('sha')).toBe(false);
      expect(params.has('baseSha')).toBe(false);
      expect(params.has('gitApplyCompatible')).toBe(false);
      return Promise.resolve(jsonResponse(diffResponse({ base_sha: 'base-sha' })));
    });

    await makeRepo().getCommitDiff({
      ref: 'preferred',
      sha: 'legacy',
      baseRef: 'preferred-base',
      baseSha: 'legacy-base',
      refIsEphemeral: false,
      baseIsEphemeral: false,
      gitApplyCompatible: true,
    });
  });

  it('sends source_ref for merge and prefers it to sourceBranch', async () => {
    mockFetch.mockImplementationOnce((_url, init) => {
      const body = JSON.parse(String(init?.body));
      expect(body.source_ref).toBe('preferred');
      expect(body).not.toHaveProperty('source_branch');
      return Promise.resolve(jsonResponse(mergeResponse({ ref: 'preferred' })));
    });

    await makeRepo().merge({
      sourceRef: 'preferred',
      sourceBranch: 'legacy',
      targetBranch: 'main',
      strategy: 'ff_only',
    });
  });

  it('sends ref for createTag and prefers it to target', async () => {
    mockFetch.mockImplementationOnce((_url, init) => {
      const body = JSON.parse(String(init?.body));
      expect(body.ref).toBe('preferred');
      expect(body).not.toHaveProperty('target');
      return Promise.resolve(
        jsonResponse({ name: 'v1', sha: 'tag-sha', message: 'created' })
      );
    });

    await makeRepo().createTag({ name: 'v1', ref: 'preferred', target: 'legacy' });
  });

  it('sends target_branch for deleteBranch and prefers it to name', async () => {
    mockFetch.mockImplementationOnce((_url, init) => {
      const body = JSON.parse(String(init?.body));
      expect(body.target_branch).toBe('preferred');
      expect(body).not.toHaveProperty('name');
      return Promise.resolve(
        jsonResponse({ target_branch: 'preferred', ephemeral: false, message: 'deleted' })
      );
    });

    await makeRepo().deleteBranch({ targetBranch: 'preferred', name: 'legacy' });
  });

  it('sends standard note query names and prefers them to aliases', async () => {
    mockFetch.mockImplementationOnce((url) => {
      const params = new URL(String(url)).searchParams;
      expect(params.get('object_ref')).toBe('preferred-object');
      expect(params.get('notes_ref')).toBe('preferred-notes');
      expect(params.has('sha')).toBe(false);
      expect(params.has('ref')).toBe(false);
      return Promise.resolve(
        jsonResponse({ sha: 'object-sha', note: 'note', ref_sha: 'notes-sha' })
      );
    });

    await makeRepo().getNote({
      objectRef: 'preferred-object',
      sha: 'legacy-object',
      notesRef: 'preferred-notes',
      ref: 'legacy-notes',
    });
  });

  it('sends standard note write fields and prefers them to aliases', async () => {
    mockFetch.mockImplementationOnce((_url, init) => {
      const body = JSON.parse(String(init?.body));
      expect(body.object_ref).toBe('preferred-object');
      expect(body.notes_ref).toBe('preferred-notes');
      expect(body.expected_notes_ref_sha).toBe('preferred-guard');
      expect(body).not.toHaveProperty('sha');
      expect(body).not.toHaveProperty('ref');
      expect(body).not.toHaveProperty('expected_ref_sha');
      return Promise.resolve(
        jsonResponse(noteResponse({ notes_ref: 'preferred-notes' }))
      );
    });

    await makeRepo().createNote({
      objectRef: 'preferred-object',
      sha: 'legacy-object',
      note: 'note',
      notesRef: 'preferred-notes',
      ref: 'legacy-notes',
      expectedNotesRefSha: 'preferred-guard',
      expectedRefSha: 'legacy-guard',
    });
  });

  it('sends standard restore fields and prefers them to aliases', async () => {
    mockFetch.mockImplementationOnce((_url, init) => {
      const body = JSON.parse(String(init?.body)).metadata;
      expect(body.base_ref).toBe('preferred-base');
      expect(body.expected_target_sha).toBe('preferred-guard');
      expect(body).not.toHaveProperty('target_commit_sha');
      expect(body).not.toHaveProperty('expected_head_sha');
      return Promise.resolve(jsonResponse(commitResponse({ target_branch: 'main' })));
    });

    await makeRepo().restoreCommit({
      targetBranch: 'main',
      baseRef: 'preferred-base',
      targetCommitSha: 'legacy-base',
      expectedTargetSha: 'preferred-guard',
      expectedHeadSha: 'legacy-guard',
      author: { name: 'Author', email: 'author@example.com' },
    });
  });

  it('sends standard fields for createCommit and prefers explicit false', async () => {
    mockFetch.mockImplementationOnce(async (_url, init) => {
      const line = (await readRequestBody(init?.body)).trim().split('\n')[0];
      const metadata = JSON.parse(line).metadata;
      expect(metadata.expected_target_sha).toBe('preferred-guard');
      expect(metadata.target_is_ephemeral).toBe(false);
      expect(metadata.base_is_ephemeral).toBe(false);
      expect(metadata).not.toHaveProperty('expected_head_sha');
      expect(metadata).not.toHaveProperty('ephemeral');
      expect(metadata).not.toHaveProperty('ephemeral_base');
      return jsonResponse(commitResponse({ target_branch: 'main' }));
    });

    await makeRepo()
      .createCommit({
        targetBranch: 'main',
        commitMessage: 'message',
        expectedTargetSha: 'preferred-guard',
        expectedHeadSha: 'legacy-guard',
        targetIsEphemeral: false,
        ephemeral: true,
        baseBranch: 'base',
        baseIsEphemeral: false,
        ephemeralBase: true,
        author: { name: 'Author', email: 'author@example.com' },
      })
      .send();
  });

  it('sends standard fields for createCommitFromDiff', async () => {
    mockFetch.mockImplementationOnce(async (_url, init) => {
      const line = (await readRequestBody(init?.body)).trim().split('\n')[0];
      const metadata = JSON.parse(line).metadata;
      expect(metadata.expected_target_sha).toBe('preferred-guard');
      expect(metadata.target_is_ephemeral).toBe(false);
      expect(metadata.base_is_ephemeral).toBe(false);
      expect(metadata).not.toHaveProperty('expected_head_sha');
      expect(metadata).not.toHaveProperty('ephemeral');
      expect(metadata).not.toHaveProperty('ephemeral_base');
      return jsonResponse(commitResponse({ target_branch: 'main' }));
    });

    await makeRepo().createCommitFromDiff({
      targetBranch: 'main',
      commitMessage: 'message',
      diff: '',
      expectedTargetSha: 'preferred-guard',
      expectedHeadSha: 'legacy-guard',
      targetIsEphemeral: false,
      ephemeral: true,
      baseBranch: 'base',
      baseIsEphemeral: false,
      ephemeralBase: true,
      author: { name: 'Author', email: 'author@example.com' },
    });
  });

  it('parses standard-only response fields and populates deprecated aliases', async () => {
    const repo = makeRepo();

    mockFetch.mockResolvedValueOnce(
      jsonResponse({
        repos: [{
          repo_id: 'repo-id',
          repo_name: 'preferred-repo',
          default_branch: 'main',
          created_at: '2026-08-28T00:00:00Z',
        }],
        has_more: false,
      })
    );
    const store = new GitStorage({ name: 'v0', token: 'header.payload.signature' });
    const repos = await store.listRepos();
    expect(repos.repos[0].repoName).toBe('preferred-repo');
    expect(repos.repos[0].url).toBe('preferred-repo');

    mockFetch.mockResolvedValueOnce(jsonResponse(diffResponse({ base_sha: 'base-sha' })));
    expect((await repo.getCommitDiff({ ref: 'main' })).baseSha).toBe('base-sha');

    mockFetch.mockResolvedValueOnce(
      jsonResponse({ target_branch: 'preferred-branch', ephemeral: false, message: 'deleted' })
    );
    const deleted = await repo.deleteBranch({ targetBranch: 'preferred-branch' });
    expect(deleted.targetBranch).toBe('preferred-branch');
    expect(deleted.name).toBe('preferred-branch');

    mockFetch.mockResolvedValueOnce(jsonResponse(commitResponse({ target_branch: 'preferred-branch' })));
    const committed = await repo.createCommitFromDiff({
      targetBranch: 'main',
      commitMessage: 'message',
      diff: '',
      author: { name: 'Author', email: 'author@example.com' },
    });
    expect(committed.refUpdate.targetBranch).toBe('preferred-branch');
    expect(committed.refUpdate.branch).toBe('preferred-branch');

    mockFetch.mockResolvedValueOnce(jsonResponse(mergeResponse({ ref: 'preferred-source' })));
    const merged = await repo.merge({ sourceRef: 'source', targetBranch: 'main', strategy: 'ff_only' });
    expect(merged.source.ref).toBe('preferred-source');
    expect(merged.source.branch).toBe('preferred-source');

    mockFetch.mockResolvedValueOnce(jsonResponse(noteResponse({ notes_ref: 'preferred-notes' })));
    const note = await repo.createNote({ objectRef: 'object', note: 'note' });
    expect(note.notesRef).toBe('preferred-notes');
    expect(note.targetRef).toBe('preferred-notes');
  });

  it('accepts deprecated-only response fields', async () => {
    const repo = makeRepo();
    const store = new GitStorage({ name: 'v0', token: 'header.payload.signature' });

    mockFetch.mockResolvedValueOnce(jsonResponse({
      repos: [{ repo_id: 'repo-id', url: 'legacy-repo', default_branch: 'main', created_at: '' }],
      has_more: false,
    }));
    expect((await store.listRepos()).repos[0].repoName).toBe('legacy-repo');

    mockFetch.mockResolvedValueOnce(jsonResponse({ name: 'legacy-branch', ephemeral: false, message: 'deleted' }));
    expect((await repo.deleteBranch({ targetBranch: 'main' })).targetBranch).toBe('legacy-branch');

    mockFetch.mockResolvedValueOnce(jsonResponse(commitResponse({ branch: 'legacy-branch' })));
    expect((await repo.createCommitFromDiff({
      targetBranch: 'main', commitMessage: 'message', diff: '',
      author: { name: 'Author', email: 'author@example.com' },
    })).refUpdate.targetBranch).toBe('legacy-branch');

    mockFetch.mockResolvedValueOnce(jsonResponse(mergeResponse({ branch: 'legacy-source' })));
    expect((await repo.merge({ sourceRef: 'source', targetBranch: 'main', strategy: 'ff_only' })).source.ref).toBe('legacy-source');

    mockFetch.mockResolvedValueOnce(jsonResponse(noteResponse({ target_ref: 'legacy-notes' })));
    expect((await repo.createNote({ objectRef: 'object', note: 'note' })).notesRef).toBe('legacy-notes');
  });

  it('prefers standard response fields when both names differ', async () => {
    const repo = makeRepo();
    const store = new GitStorage({ name: 'v0', token: 'header.payload.signature' });

    mockFetch.mockResolvedValueOnce(jsonResponse({
      repos: [{ repo_id: 'repo-id', repo_name: 'preferred-repo', url: 'legacy-repo', default_branch: 'main', created_at: '' }],
      has_more: false,
    }));
    const listed = (await store.listRepos()).repos[0];
    expect([listed.repoName, listed.url]).toEqual(['preferred-repo', 'preferred-repo']);

    mockFetch.mockResolvedValueOnce(jsonResponse({ target_branch: 'preferred-branch', name: 'legacy-branch', ephemeral: false, message: 'deleted' }));
    const deleted = await repo.deleteBranch({ targetBranch: 'main' });
    expect([deleted.targetBranch, deleted.name]).toEqual(['preferred-branch', 'preferred-branch']);

    mockFetch.mockResolvedValueOnce(jsonResponse(commitResponse({ target_branch: 'preferred-branch', branch: 'legacy-branch' })));
    const committed = await repo.createCommitFromDiff({
      targetBranch: 'main', commitMessage: 'message', diff: '',
      author: { name: 'Author', email: 'author@example.com' },
    });
    expect([committed.refUpdate.targetBranch, committed.refUpdate.branch]).toEqual(['preferred-branch', 'preferred-branch']);

    mockFetch.mockResolvedValueOnce(jsonResponse(mergeResponse({ ref: 'preferred-source', branch: 'legacy-source' })));
    const merged = await repo.merge({ sourceRef: 'source', targetBranch: 'main', strategy: 'ff_only' });
    expect([merged.source.ref, merged.source.branch]).toEqual(['preferred-source', 'preferred-source']);

    mockFetch.mockResolvedValueOnce(jsonResponse(noteResponse({ notes_ref: 'preferred-notes', target_ref: 'legacy-notes' })));
    const note = await repo.createNote({ objectRef: 'object', note: 'note' });
    expect([note.notesRef, note.targetRef]).toEqual(['preferred-notes', 'preferred-notes']);
  });
});

File: packages/code-storage-go/standard_vocabulary_test.go
package storage

import (
	"bytes"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"net/url"
	"strings"
	"testing"
)

type capturedStandardRequest struct {
	query url.Values
	body  map[string]interface{}
}

func TestStandardVocabularyRequests(t *testing.T) {
	captured := map[string][]capturedStandardRequest{}
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		request := capturedStandardRequest{query: r.URL.Query()}
		body, _ := io.ReadAll(r.Body)
		if len(bytes.TrimSpace(body)) > 0 {
			if r.URL.Path == "/api/v1/repos/commit-pack" || r.URL.Path == "/api/v1/repos/diff-commit" {
				var envelope map[string]map[string]interface{}
				_ = json.Unmarshal(bytes.Split(body, []byte("\n"))[0], &envelope)
				request.body = envelope["metadata"]
			} else {
				_ = json.Unmarshal(body, &request.body)
			}
		}
		key := r.Method + " " + r.URL.Path
		captured[key] = append(captured[key], request)
		w.Header().Set("Content-Type", "application/json")

		switch key {
		case "GET /api/v1/repos/commits":
			_, _ = w.Write([]byte(`{"commits":[],"has_more":false}`))
		case "GET /api/v1/repos/commit":
			_, _ = w.Write([]byte(`{"commit":{"sha":"sha","parent_shas":[],"message":"message","author_name":"A","author_email":"a@example.com","committer_name":"A","committer_email":"a@example.com","date":"2026-08-28T00:00:00Z"}}`))
		case "GET /api/v1/repos/diff":
			_, _ = w.Write([]byte(`{"sha":"sha","base_sha":"base","stats":{},"files":[],"filtered_files":[]}`))
		case "GET /api/v1/repos/notes":
			_, _ = w.Write([]byte(`{"sha":"sha","note":"note","ref_sha":"notes-sha"}`))
		case "POST /api/v1/repos/merge":
			_, _ = w.Write([]byte(`{"result":"fast_forward","commit_sha":"sha","tree_sha":"tree","source":{"ref":"source","ephemeral":false,"sha":"sha"},"target":{"branch":"main","ephemeral":false,"old_sha":"old","new_sha":"sha"},"promoted_commits":1}`))
		case "POST /api/v1/repos/tags":
			_, _ = w.Write([]byte(`{"name":"v1","sha":"sha","message":"created"}`))
		case "DELETE /api/v1/repos/branches":
			_, _ = w.Write([]byte(`{"target_branch":"branch","message":"deleted","ephemeral":false}`))
		case "POST /api/v1/repos/notes", "DELETE /api/v1/repos/notes":
			_, _ = w.Write([]byte(`{"sha":"sha","notes_ref":"refs/notes/reviews","new_ref_sha":"notes-sha","result":{"success":true,"status":"ok"}}`))
		case "POST /api/v1/repos/restore-commit":
			_, _ = w.Write([]byte(`{"commit":{"commit_sha":"sha","tree_sha":"tree","target_branch":"main","pack_bytes":1},"result":{"target_branch":"main","old_sha":"old","new_sha":"sha","success":true,"status":"ok"}}`))
		case "POST /api/v1/repos/commit-pack", "POST /api/v1/repos/diff-commit":
			_, _ = w.Write([]byte(`{"commit":{"commit_sha":"sha","tree_sha":"tree","target_branch":"main","pack_bytes":1,"blob_count":0},"result":{"target_branch":"main","old_sha":"old","new_sha":"sha","success":true,"status":"ok"}}`))
		default:
			t.Fatalf("unexpected request: %s", key)
		}
	}))
	defer server.Close()

	client, err := NewClient(Options{Name: "acme", Key: testKey, APIBaseURL: server.URL})
	if err != nil {
		t.Fatalf("client error: %v", err)
	}
	repo := &Repo{ID: "repo", DefaultBranch: "main", client: client}

	_, _ = repo.ListCommits(nil, ListCommitsOptions{Ref: "preferred", Branch: "legacy"})
	_, _ = repo.GetCommit(nil, GetCommitOptions{Ref: "preferred", SHA: "legacy"})
	_, _ = repo.GetCommitDiff(nil, GetCommitDiffOptions{
		Ref: "preferred", SHA: "legacy", BaseRef: "preferred-base", BaseSHA: "legacy-base",
		RefIsEphemeral: boolPtr(false), BaseIsEphemeral: boolPtr(false), GitApplyCompatible: true,
	})
	_, _ = repo.GetNote(nil, GetNoteOptions{
		ObjectRef: "preferred", SHA: "legacy", NotesRef: "preferred-notes", Ref: "legacy-notes",
	})
	_, _ = repo.Merge(nil, MergeOptions{
		SourceRef: "preferred", SourceBranch: "legacy", TargetBranch: "main", Strategy: MergeStrategyFFOnly,
	})
	_, _ = repo.CreateTag(nil, CreateTagOptions{Name: "v1", Ref: "preferred", Target: "legacy"})
	_, _ = repo.DeleteBranch(nil, DeleteBranchOptions{TargetBranch: "preferred", Name: "legacy"})
	_, _ = repo.CreateNote(nil, CreateNoteOptions{
		ObjectRef: "preferred", SHA: "legacy", Note: "note", NotesRef: "preferred-notes", Ref: "legacy-notes",
		ExpectedNotesRefSHA: "preferred-guard", ExpectedRefSHA: "legacy-guard",
	})
	_, _ = repo.AppendNote(nil, AppendNoteOptions{ObjectRef: "object", Note: "note"})
	_, _ = repo.DeleteNote(nil, DeleteNoteOptions{ObjectRef: "object"})
	_, _ = repo.RestoreCommit(nil, RestoreCommitOptions{
		TargetBranch: "main", BaseRef: "preferred", TargetCommitSHA: "legacy",
		ExpectedTargetSHA: "preferred-guard", ExpectedHeadSHA: "legacy-guard",
		Author: CommitSignature{Name: "Author", Email: "author@example.com"},
	})

	preferredFalse := false
	builder, _ := repo.CreateCommit(CommitOptions{
		TargetBranch: "main", CommitMessage: "message", Author: CommitSignature{Name: "Author", Email: "author@example.com"},
		ExpectedTargetSHA: "preferred-guard", ExpectedHeadSHA: "legacy-guard",
		BaseBranch: "base", TargetIsEphemeral: &preferredFalse, Ephemeral: true,
		BaseIsEphemeral: &preferredFalse, EphemeralBase: true,
	})
	_, _ = builder.Send(nil)
	_, _ = repo.CreateCommitFromDiff(nil, CommitFromDiffOptions{
		TargetBranch: "main", CommitMessage: "message", Author: CommitSignature{Name: "Author", Email: "author@example.com"},
		Diff: strings.NewReader("diff"), ExpectedTargetSHA: "preferred-guard", ExpectedHeadSHA: "legacy-guard",
		BaseBranch: "base", TargetIsEphemeral: &preferredFalse, Ephemeral: true,
		BaseIsEphemeral: &preferredFalse, EphemeralBase: true,
	})

	assertQuery := func(key string, expected url.Values) {
		t.Helper()
		if got := captured[key][0].query; got.Encode() != expected.Encode() {
			t.Fatalf("%s query = %v, want %v", key, got, expected)
		}
	}
	assertQuery("GET /api/v1/repos/commits", url.Values{"ref": {"preferred"}})
	assertQuery("GET /api/v1/repos/commit", url.Values{"ref": {"preferred"}})
	assertQuery("GET /api/v1/repos/diff", url.Values{
		"ref": {"preferred"}, "base_ref": {"preferred-base"}, "ref_is_ephemeral": {"false"},
		"base_is_ephemeral": {"false"}, "git_apply_compatible": {"true"},
	})
	assertQuery("GET /api/v1/repos/notes", url.Values{"object_ref": {"preferred"}, "notes_ref": {"preferred-notes"}})

	assertBodyField(t, captured["POST /api/v1/repos/merge"][0].body, "source_ref", "preferred", "source_branch")
	assertBodyField(t, captured["POST /api/v1/repos/tags"][0].body, "ref", "preferred", "target")
	assertBodyField(t, captured["DELETE /api/v1/repos/branches"][0].body, "target_branch", "preferred", "name")
	assertBodyField(t, captured["POST /api/v1/repos/notes"][0].body, "object_ref", "preferred", "sha")
	assertBodyField(t, captured["POST /api/v1/repos/notes"][0].body, "notes_ref", "preferred-notes", "ref")
	assertBodyField(t, captured["POST /api/v1/repos/notes"][0].body, "expected_notes_ref_sha", "preferred-guard", "expected_ref_sha")
	assertBodyField(t, captured["POST /api/v1/repos/notes"][1].body, "object_ref", "object", "sha")
	assertBodyField(t, captured["DELETE /api/v1/repos/notes"][0].body, "object_ref", "object", "sha")
	assertBodyField(t, captured["POST /api/v1/repos/restore-commit"][0].body["metadata"].(map[string]interface{}), "base_ref", "preferred", "target_commit_sha")
	assertBodyField(t, captured["POST /api/v1/repos/restore-commit"][0].body["metadata"].(map[string]interface{}), "expected_target_sha", "preferred-guard", "expected_head_sha")

	for _, key := range []string{"POST /api/v1/repos/commit-pack", "POST /api/v1/repos/diff-commit"} {
		body := captured[key][0].body
		assertBodyField(t, body, "expected_target_sha", "preferred-guard", "expected_head_sha")
		if body["target_is_ephemeral"] != false || body["base_is_ephemeral"] != false {
			t.Fatalf("%s did not preserve preferred false flags: %#v", key, body)
		}
		if _, ok := body["ephemeral"]; ok {
			t.Fatalf("%s sent deprecated ephemeral", key)
		}
		if _, ok := body["ephemeral_base"]; ok {
			t.Fatalf("%s sent deprecated ephemeral_base", key)
		}
	}
}

func assertBodyField(t *testing.T, body map[string]interface{}, preferred string, value interface{}, deprecated string) {
	t.Helper()
	if body[preferred] != value {
		t.Fatalf("%s = %#v, want %#v", preferred, body[preferred], value)
	}
	if _, ok := body[deprecated]; ok {
		t.Fatalf("request sent deprecated field %s", deprecated)
	}
}

func TestStandardVocabularyCommitResponseAliases(t *testing.T) {
	for _, test := range []struct {
		name         string
		targetBranch string
		branch       string
		want         string
	}{
		{name: "standard only", targetBranch: "standard", want: "standard"},
		{name: "deprecated only", branch: "legacy", want: "legacy"},
		{name: "standard wins", targetBranch: "standard", branch: "legacy", want: "standard"},
	} {
		t.Run(test.name, func(t *testing.T) {
			var ack commitPackAck
			ack.Commit.CommitSHA = "sha"
			ack.Commit.TreeSHA = "tree"
			ack.Commit.TargetBranch = "main"
			if test.targetBranch != "" {
				ack.Result.TargetBranch = &test.targetBranch
			}
			ack.Result.Branch = test.branch
			ack.Result.Success = true
			result, err := buildCommitResult(ack)
			if err != nil {
				t.Fatalf("build result: %v", err)
			}
			if result.RefUpdate.TargetBranch != test.want || result.RefUpdate.Branch != test.want {
				t.Fatalf("unexpected aliases: %#v", result.RefUpdate)
			}

			var restore restoreCommitAck
			restore.Commit.CommitSHA = "sha"
			restore.Commit.TreeSHA = "tree"
			restore.Commit.TargetBranch = "main"
			restore.Result.TargetBranch = ack.Result.TargetBranch
			restore.Result.Branch = test.branch
			restore.Result.Success = true
			restored, err := buildRestoreCommitResult(restore)
			if err != nil {
				t.Fatalf("build restore result: %v", err)
			}
			if restored.RefUpdate.TargetBranch != test.want || restored.RefUpdate.Branch != test.want {
				t.Fatalf("unexpected restore aliases: %#v", restored.RefUpdate)
			}
		})
	}
}

func TestStandardVocabularyResponseAliases(t *testing.T) {
	for _, test := range []struct {
		name              string
		includeStandard   bool
		includeDeprecated bool
		want              string
	}{
		{name: "standard only", includeStandard: true, want: "standard"},
		{name: "deprecated only", includeDeprecated: true, want: "legacy"},
		{name: "standard wins", includeStandard: true, includeDeprecated: true, want: "standard"},
	} {
		t.Run(test.name, func(t *testing.T) {
			server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				w.Header().Set("Content-Type", "application/json")
				response := map[string]interface{}{}
				switch r.Method + " " + r.URL.Path {
				case "GET /api/v1/repos":
					repo := map[string]interface{}{"repo_id": "repo", "default_branch": "main", "created_at": ""}
					setVocabularyFields(repo, test.includeStandard, test.includeDeprecated, "repo_name", "url")
					response = map[string]interface{}{"repos": []interface{}{repo}, "has_more": false}
				case "GET /api/v1/repos/diff":
					response = map[string]interface{}{"sha": "sha", "base_sha": "base", "stats": map[string]interface{}{}, "files": []interface{}{}, "filtered_files": []interface{}{}}
				case "DELETE /api/v1/repos/branches":
					response = map[string]interface{}{"message": "deleted", "ephemeral": false}
					setVocabularyFields(response, test.includeStandard, test.includeDeprecated, "target_branch", "name")
				case "POST /api/v1/repos/merge":
					source := map[string]interface{}{"ephemeral": false, "sha": "sha"}
					setVocabularyFields(source, test.includeStandard, test.includeDeprecated, "ref", "branch")
					response = map[string]interface{}{
						"result": "fast_forward", "commit_sha": "sha", "tree_sha": "tree", "source": source,
						"target": map[string]interface{}{"branch": "main", "ephemeral": false, "old_sha": "old", "new_sha": "sha"}, "promoted_commits": 1,
					}
				case "POST /api/v1/repos/notes":
					response = map[string]interface{}{"sha": "sha", "new_ref_sha": "notes-sha", "result": map[string]interface{}{"success": true, "status": "ok"}}
					setVocabularyFields(response, test.includeStandard, test.includeDeprecated, "notes_ref", "target_ref")
				default:
					t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
				}
				if err := json.NewEncoder(w).Encode(response); err != nil {
					t.Fatalf("encode response: %v", err)
				}
			}))
			defer server.Close()

			client, err := NewClient(Options{Name: "acme", Key: testKey, APIBaseURL: server.URL})
			if err != nil {
				t.Fatalf("client error: %v", err)
			}
			repo := &Repo{ID: "repo", DefaultBranch: "main", client: client}

			listed, err := client.ListRepos(nil, ListReposOptions{})
			if err != nil {
				t.Fatalf("list repos: %v", err)
			}
			if listed.Repos[0].RepoName != test.want || listed.Repos[0].URL != test.want {
				t.Fatalf("unexpected repo aliases: %#v", listed.Repos[0])
			}
			diff, err := repo.GetCommitDiff(nil, GetCommitDiffOptions{Ref: "main"})
			if err != nil || diff.BaseSHA != "base" {
				t.Fatalf("unexpected diff result: %#v, %v", diff, err)
			}
			deleted, err := repo.DeleteBranch(nil, DeleteBranchOptions{TargetBranch: "branch"})
			if err != nil || deleted.TargetBranch != test.want || deleted.Name != test.want {
				t.Fatalf("unexpected delete aliases: %#v, %v", deleted, err)
			}
			merged, err := repo.Merge(nil, MergeOptions{SourceRef: "source", TargetBranch: "main", Strategy: MergeStrategyFFOnly})
			if err != nil || merged.Source.Ref != test.want || merged.Source.Branch != test.want {
				t.Fatalf("unexpected merge aliases: %#v, %v", merged.Source, err)
			}
			note, err := repo.CreateNote(nil, CreateNoteOptions{ObjectRef: "object", Note: "note"})
			if err != nil || note.NotesRef != test.want || note.TargetRef != test.want {
				t.Fatalf("unexpected note aliases: %#v, %v", note, err)
			}
		})
	}
}

func setVocabularyFields(body map[string]interface{}, standard bool, deprecated bool, standardName string, deprecatedName string) {
	if standard {
		body[standardName] = "standard"
	}
	if deprecated {
		body[deprecatedName] = "legacy"
	}
}
